### PR TITLE
RFC: Use JL_LLVM_VERSION instead of LLVMxx

### LINKED ISSUE #9860: Replace `LLVM_34` and friends by `LLVM_VERSION`
Julia's LLVM interface code contains many constructs of the kind
```
#ifdef LLVM33
            #define LLVM_FP(a,b) APFloat(a,b)
#else
            #define LLVM_FP(a,b) APFloat(b,true)
#endif
```
which handle the API changes in LLVM between different versions. The macros `LLVM33` have a somewhat counter-intuitive meaning in that they mean "version x OR LATER". In the case above, the `#else` branch thus only triggers for LLVM 3.2 and earlier. This is confusing.

I suggest to replace the various `LLVM*` macros by a single macro `LLVM_VERSION, and check this explicitly, as in `#if LLVM_VERSION >= 34` (or similar). This is easier to understand, and also reduces the number of macros that need to be defined.

Alternatively, we can rename `LLVM33` etc. to `LLVM33_OR_LATER` etc.

### PR DESCRIPTION
**Pull request by ararslan**
Fixes #9860.

As noted in the linked issue, the names `LLVM37`, `LLVM34`, etc. aren't immediately descriptive because they don't refer to specific versions but rather to ranges of versions. This implements the proposed fix of having a single `LLVM_VERSION` macro that holds the LLVM version as a two-digit number and instead of checking whether `LLVMxx` is defined, we check whether `LLVM_VERSION >= xx`. Similarly, checking whether `LLVMxx` is _not_ defined equates to `LLVM_VERSION < xx`.

This works locally, so fingers crossed that it works everywhere else...

### CODE REVIEW

**Comment by yuyichao** [comment]
There was some discussion of whether there should be a 3.10 version. Just to be safe, I think we can use something like `0x030701`.

File: src/llvm-version.h [hunk 1]
```diff
@@ -2,53 +2,49 @@
...
+        #elif LLVM_VERSION_MINOR == 6
+            #define LLVM_VERSION 36
+        #elif LLVM_VERSION_MINOR == 5
+            #define LLVM_VERSION 35
+        #elif LLVM_VERSION_MINOR == 4
+            #define LLVM_VERSION 34
+        #elif LLVM_VERSION_MINOR == 3
+            #define LLVM_VERSION 33
+        #endif
+    #endif
```
Comment (ararslan) [inline src/llvm-version.h]: Constructing `LLVM_VERSION` can be done more succinctly as
```c
#define CONCAT(a, b) a ## b
#define GETVERS(maj, min) CONCAT(maj, min)
#define LLVM_VERSION GETVERS(LLVM_VERSION_MAJOR, LLVM_VERSION_MINOR)
```
But as @yuyichao noted, this is the kind of thing that one gets stabbed for if put into production.

**Comment by ararslan** [comment]
You mean check for LLVM 3.10 by doing `LLVM_VERSION >= 0x030701`?

**Comment by yuyichao** [comment]
No, I mean `#define JL_LLVM_VERSION (LLVM_VERSION_MAJOR * 0x10000 + LLVM_VERSION_MINOR * 0x100 + LLVM_VERSION_PATCH)`

**Comment by yuyichao** [comment]
`100` instead of `0x100` is also fine, and can possibly reduce typo (it's probably more likely to miss a `0x` than adding one)

**Comment by yuyichao** [comment]
Anyther way is to use something like the glibc trick
https://github.com/JuliaLang/julia/blob/3e00217d69fa3f59dfd1ee0a6fd04deccec0dd5b/src/threading.c#L153

**Comment by ararslan** [comment]
So something like
```c
#if _LLVM_VERSION_GE(3, 10)
```
where `_LLVM_VERSION_GE(x, y)` checks that the version is at least `x.y`?

**Comment by ararslan** [comment]
I get segfaults when I do `#define JL_LLVM_VERSION (LLVM_VERSION_MAJOR * 0x10000 + LLVM_VERSION_MINOR * 0x100 + LLVM_VERSION_PATCH)`

**Comment by yuyichao** [comment]
Yes.... Maybe use `JL_` instead of `_` as prefix. IIRC, there is one place we check patch version in this file. Too bad msvc doesn't like variadic macros otherwise we can use the same macro to check version with an optional patch version....

**Comment by yuyichao** [comment]
> I get segfaults when I do

Full patch??

**Comment by ararslan** [comment]
What do you mean full patch? This is what I was doing to test:
```c
#define LLVM_VERSION_MAJOR 3
#define LLVM_VERSION_MINOR 10
#define LLVM_VERSION_PATCH 1

#define JL_LLVM_VERSION (LLVM_VERSION_MAJOR * 0x10000 + LLVM_VERSION_MINOR * 0x100 + LLVM_VERSION_PATCH)

int main() {
    int x = JL_LLVM_VERSION;
    printf("%s", x);
    return 0;
}
```

**Comment by yuyichao** [comment]
```c
    int x = JL_LLVM_VERSION;
    printf("%s", x);
    return 0;
```

Why should this work at all?

**Comment by ararslan** [comment]
¯\\\_(ツ)\_/¯  Because I'm not good at C

**Comment by yuyichao** [comment]
> Because I'm not good at C

A good advice for C is to turn on `-Wall -Wextra` and fix all the warning.

**Comment by ararslan** [comment]
Oh, heh, whoops. Format string...

So for major version 3, minor 10, patch 1, I get 199169. How would you check the version based on that number?

**Comment by yuyichao** [comment]
`0x31001`

**Comment by ararslan** [comment]
Do patches ever have more than one digit?

**Comment by yuyichao** [comment]
> Do patches ever have more than two digits?

The nope is that neither patch and minor do. Actually `100` is better than `0x100` sin `3.10` is actually `0x30a00` ........

**Comment by ararslan** [comment]
Okay, changed to use the `0x` version you suggested.

**Comment by yuyichao** [comment]
I actually mean the `100` version is better ....

File: src/llvm-version.h [hunk 2]
```diff
@@ -2,53 +2,37 @@
...
 
-// We enable ORCJIT only if we have our custom patches
-#ifndef SYSTEM_LLVM
-#define USE_ORCJIT
-#endif
+    #define LLVM_VERSION (LLVM_VERSION_MAJOR * 0x10000 + LLVM_VERSION_MINOR * 0x100 \
+                        + LLVM_VERSION_PATCH)
 
-#endif
+    #if LLVM_VERSION < 0x33000
```
Comment (yuyichao) [inline src/llvm-version.h]: I think thse bounds are wrong. E.g. after switching to `100`, this should be `30300`

File: src/llvm-version.h [hunk 3]
```diff
@@ -2,53 +2,37 @@
...
-#define LLVM38 1
-#define USE_ORCJIT
-#endif
-
-#if defined(LLVM40) || (defined(LLVM_VERSION_MAJOR) && LLVM_VERSION_MAJOR == 3 && LLVM_VERSION_MINOR >= 7)
-#define LLVM37 1
+// The version LLVM version used, LLVM_VERSION, is represented as a hexadecimal
+// string: 0xABBCC where A is the major version, B is minor, and C is patch. So
+// for example, LLVM 3.7.0 is 0x30700.
+#ifdef LLVM_VERSION_MAJOR
```
Comment (yuyichao) [inline src/llvm-version.h]: This should be removed.

File: src/llvm-version.h [hunk 4]
```diff
@@ -2,53 +2,37 @@
...
 
-// We enable ORCJIT only if we have our custom patches
-#ifndef SYSTEM_LLVM
-#define USE_ORCJIT
-#endif
+    #define LLVM_VERSION (LLVM_VERSION_MAJOR * 0x10000 + LLVM_VERSION_MINOR * 0x100 \
```
Comment (yuyichao) [inline src/llvm-version.h]: And I think we should use a `JL_` prefix. This name is too close to what LLVM might decide to define some day.

File: src/llvm-version.h (hunk 3 above)
Comment (ararslan) [inline src/llvm-version.h]: You mean the comment or the check for whether the version is defined?

Comment (yuyichao) [inline src/llvm-version.h]: The check of whether the version is defined. The comment is great!

File: src/llvm-version.h (hunk 4 above)
Comment (yuyichao) [inline src/llvm-version.h]: This is still not addressed yet.

**Comment by yuyichao** [comment]
LGTM other than the missing `JL_` prefix.

**Comment by ararslan** [comment]
Okay, added the `JL_` prefix. I had meant to add it when you had first mentioned it but forgot. Thanks so much for all of your help and patience!

File: src/codegen_internal.h [hunk 5]
```diff
@@ -1,6 +1,6 @@
 // This file is a part of Julia. License is MIT: http://julialang.org/license
 
-#if defined(LLVM38) && !defined(LLVM37)
+#if defined(JL_LLVM_VERSION) && JL_LLVM_VERSION >= 30800
```
Comment (yuyichao) [inline src/codegen_internal.h]: Is the `defined(JL_LLVM_VERSION)` needed here?

Comment (ararslan) [inline src/codegen_internal.h]: No, that's not needed. I'll take it out.

Comment (yuyichao) [inline src/codegen_internal.h]: IMHO, we shouldn't check if `JL_LLVM_VERSION` is defined. This can catch errors of missing include earlier rather than silently ignoring the error.

**Comment by yuyichao** [comment]
I looked through the diff and I think the translation is correct. Since the diff is so long, it would be better if someone else could have a look too to make sure we aren't silently disabling features on newer version of llvm.

File: src/jitlayers.cpp [hunk 6]
```diff
@@ -103,7 +103,7 @@ void addOptimizationPasses(PassManager *PM)
 #endif
 
 #if defined(JL_ASAN_ENABLED)
-#   if defined(LLVM37) && !defined(LLVM38)
+#   if JL_LLVM_VERSION == 30700
```
Comment (vtjnash) [inline src/jitlayers.cpp]: `#if JL_LLVM_VERSION >= 30700 && JL_LLVM_VERSION < 30800`

File: src/llvm-version.h [hunk 7]
```diff
@@ -2,53 +2,34 @@
...
-#endif
+// The LLVM version used, JL_LLVM_VERSION, is represented as a 5-digit integer
+// of the form ABBCC, where A is the major version, B is minor, and C is patch.
+// So for example, LLVM 3.7.0 is 30700.
+#define JL_LLVM_VERSION (LLVM_VERSION_MAJOR * 10000 + LLVM_VERSION_MINOR * 100 \
+                    + LLVM_VERSION_PATCH)
 
-#if defined(LLVM40) || (defined(LLVM_VERSION_MAJOR) && LLVM_VERSION_MAJOR == 3 && LLVM_VERSION_MINOR >= 9)
-#define LLVM39 1
+#if JL_LLVM_VERSION < 30300
```
Comment (vtjnash) [inline src/llvm-version.h]: `#if JL_LLVM_VERSION != 30300 && JL_LLVM_VERSION < 30701`

File: src/llvm-version.h [hunk 8]
```diff
@@ -2,53 +2,34 @@
...
-
-#endif
-
-#if defined(LLVM40) || (defined(LLVM_VERSION_MAJOR) && LLVM_VERSION_MAJOR == 3 && LLVM_VERSION_MINOR >= 6)
-#define LLVM36 1
-#endif
-
-#if defined(LLVM40) || (defined(LLVM_VERSION_MAJOR) && LLVM_VERSION_MAJOR == 3 && LLVM_VERSION_MINOR >= 5)
-#define LLVM35 1
+#if JL_LLVM_VERSION >= 30700 && !defined(SYSTEM_LLVM)
```
Comment (vtjnash) [inline src/llvm-version.h]: `#elif`

File: src/llvm-version.h (hunk 7 above)
Comment (yuyichao) [inline src/llvm-version.h]: So dropping (source code) support for 3.4-3.6? (after effectively dropped it for a long time).

**Comment by tkelman** [comment]
this could make cherry picking kind of hard, so since this is mostly a cleanup I think it would be better to wait until things are a little quieter to merge this

**Comment by yuyichao** [comment]
True. On the other hand, the conflicts should be mostly trivial and it is a little annoy to rebase this/review this multiple times.

File: src/codegen_internal.h [hunk 9]
```diff
@@ -1,6 +1,6 @@
 // This file is a part of Julia. License is MIT: http://julialang.org/license
 
-#if defined(LLVM38) && !defined(LLVM37)
+#if JL_LLVM_VERSION >= 30800
```
Comment (tkelman) [inline src/codegen_internal.h]: how would the old version of this ever happen?

File: src/debuginfo.cpp [hunk 10]
```diff
@@ -34,12 +34,12 @@
...
 #include <llvm/Object/COFF.h>
-#   ifdef LLVM37
+#   if JL_LLVM_VERSION >= 30700
 #       include <llvm/Object/ELFObjectFile.h>
 #   endif
 #endif
 
-#if defined(USE_MCJIT) && !defined(LLVM36) && defined(_OS_DARWIN_)
+#if defined(USE_MCJIT) && JL_LLVM_VERSION < 30600 && defined(_OS_DARWIN_)
 #include "../deps/llvm-3.5.0/lib/ExecutionEngine/MCJIT/MCJIT.h"
```
Comment (tkelman) [inline src/debuginfo.cpp]: this path is hardcoded and wrong, we don't really support 3.4 or 3.5 anyway

File: src/llvm-version.h (hunk 7 above)
Comment (tkelman) [inline src/llvm-version.h]: and 3.7.0 ?

File: src/codegen_internal.h (hunk 9 above)
Comment (yuyichao) [inline src/codegen_internal.h]: The function is also declared in `jitlayers.cpp`. Doing it here might have some issue due to missing type declaration/missing headers.

Comment (ararslan) [inline src/codegen_internal.h]: @tkelman If 3.8 is the lowest version. So it's at least 3.8 and not 3.7 or lower.

File: src/debuginfo.cpp (hunk 10 above)
Comment (ararslan) [inline src/debuginfo.cpp]: Should I bump the error check to say that anything less than 3.6 isn't supported? Currently it only does that for <3.3.

File: src/codegen_internal.h (hunk 9 above)
Comment (ararslan) [inline src/codegen_internal.h]: @yuyichao Should the function declaration here be removed then?

Comment (yuyichao) [inline src/codegen_internal.h]: The condition is actually `USE_ORCJIT && JL_LLVM_VERSION <= 30800` (roughly 3.7 <= llvm_ver < 3.8). It is better to remove the declaration in `jitlayers.cpp` instead and keep this one with the condition fixed.

File: src/debuginfo.cpp (hunk 10 above)
Comment (yuyichao) [inline src/debuginfo.cpp]: The new check below is already doing that so the error message should be updated too

File: src/codegen_internal.h (hunk 9 above)
Comment (ararslan) [inline src/codegen_internal.h]: I don't see how that condition is equivalent to what's currently there. Or are you proposing a change?

Comment (yuyichao) [inline src/codegen_internal.h]: The condition there before was `false`, the current condition is wrong. The correct one is in `jitlayers.cpp` and is the same with the one I gave above.

File: src/llvm-version.h [hunk 11]
```diff
@@ -8,8 +8,8 @@
 #define JL_LLVM_VERSION (LLVM_VERSION_MAJOR * 10000 + LLVM_VERSION_MINOR * 100 \
                         + LLVM_VERSION_PATCH)
 
-#if JL_LLVM_VERSION != 30300 && JL_LLVM_VERSION < 30701
-    #error LLVM versions < 3.3 are not supported by Julia
+#if JL_LLVM_VERSION < 30600
```
Comment (yuyichao) [inline src/llvm-version.h]: I think @vtjnash meant that we are supporting LLVM 3.3 and >= 3.7.1.

Comment (yuyichao) [inline src/llvm-version.h]: At least that's what we currently support (although LLVM 3.3 doesn't pass all tests anymore IIRC). We'll probably drop 3.3 support and require at least LLVM 3.7.1 (or even LLVM 3.8?) during the 0.6 cycle but that should probably be another PR (since this PR is already pretty big as it is now.)

File: src/codegen_internal.h (hunk 9 above)
Comment (ararslan) [inline src/codegen_internal.h]: Looks like it fails if I remove the part in jitlayers.cpp.

Comment (yuyichao) [inline src/codegen_internal.h]: Move to `jitlayers.h`

File: src/debuginfo.cpp (hunk 10 above)
Comment (tkelman) [inline src/debuginfo.cpp]: we support 3.3 at least for building against it. because of the backporting issues this PR will cause, ideally it wouldn't change any behavior and I could backport it. in a separate PR we could remove all 3.3 support and a lot of these ifdefs

File: src/codegen_internal.h (hunk 9 above)
Comment (tkelman) [inline src/codegen_internal.h]: `defined(LLVM38) && !defined(LLVM37)` would never happen since `LLVM37` meant "3.7 or newer"

Comment (yuyichao) [inline src/codegen_internal.h]: Which is why I said,

> The condition there before was false

This was added before the JIT and codegen are split and therefore before `jitlayers.h` exists. This is needed only by `jitlayers.cpp`, which doesn't include `codegen_internal.h`, for LLVM 3.7.x which is why it should be moved to `jitlayers.h`.

File: src/codegen_internal.h [hunk 12]
```diff
@@ -1,6 +1,6 @@
 // This file is a part of Julia. License is MIT: http://julialang.org/license
 
-#if defined(LLVM38) && !defined(LLVM37)
+#if defined(USE_ORCJIT) && JL_LLVM_VERSION <= 30800
```
Comment (tkelman) [inline src/codegen_internal.h]: since the old version of this could never be true, probably best to just delete these few lines

**Comment by tkelman** [comment]
(I found this a little easier to review locally after
```sh
curl -L https://github.com/JuliaLang/julia/pull/17826.patch | git apply
for i in $(seq 3 9); do
  sed -i -e "s/JL_LLVM_VERSION >= 30${i}00/defined(LLVM3$i)/g" \
    -e "s/JL_LLVM_VERSION < 30${i}00/\!defined(LLVM3$i)/g" \
    -e "s/#if defined(LLVM3$i)/#ifdef LLVM3$i/g" \
    -e "s/#if \!defined(LLVM3$i)/#ifndef LLVM3$i/g" src/*.cpp src/*.h
done
```
which gives a shorter diff)

File: src/codegen_internal.h (hunk 12 above)
Comment (ararslan) [inline src/codegen_internal.h]: I think @yuyichao specifically requested this change: https://github.com/JuliaLang/julia/pull/17826#discussion_r73792427

**Comment by ararslan** [comment]
Is there anything else this needs before it's good to go?

File: src/debuginfo.cpp [hunk 13]
```diff
@@ -59,7 +59,7 @@ using namespace llvm;
 #include <cstdio>
 #include <cassert>
 
-#if defined(LLVM35) && !defined(LLVM36)
+#if JL_LLVM_VERSION == 30500
```
Comment (yuyichao) [inline src/debuginfo.cpp]: This should in principle still be a range since LLVM 3.5.1 does exist. Not sure if it actually matters though since we might not compile there anyway....

**Comment by yuyichao** [comment]
LGTM other than the very minor comment about. I think we should merge this now. It's unlikely that we can trivially backport all codegen changes anymore and the conflict introduced by this change should be easy to resolve.

If everyone is fine with this I'd like to merge this tomorrow (with the comment above fixed or not) and rebase https://github.com/JuliaLang/julia/pull/18519 on top of this.

File: src/debuginfo.cpp (hunk 13 above)
Comment (ararslan) [inline src/debuginfo.cpp]: Fixed, thanks!

**Comment by yuyichao** [comment]
I think there was a conflict before #18519 was merged........ So I guess this needs to be rebased again (hopefully the last time).

**Comment by ararslan** [comment]
Rebased and conflicts resolved. Thanks for the heads up!

**Comment by ararslan** [comment]
Reviewers should probably pay extra careful attention to future PRs that modify or introduce LLVM version conditioning in the C preprocessor, since `#ifdef LLVMxx` will _always_ be false now.

**Comment by tkelman** [comment]
should we make those defines result in compile-time errors now? maybe just for a little while?

**Comment by ararslan** [comment]
That seems like a good idea to me. How would you propose we go about that? I can submit a PR.

**Comment by yuyichao** [comment]
I don't think there's a way to do it in C++ code (a grep on CI could do that though). It shouldn't be a huge issue since hopefully PR's that uses those are already tested on newer LLVM versions.
